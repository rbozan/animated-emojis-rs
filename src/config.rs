use std::{collections::HashMap, path::PathBuf, str::FromStr};

use config::Environment;
use once_cell::sync::Lazy;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct Config {
    #[serde(rename = "noto_emoji_metadata_path")]
    pub noto_emoji_metadata_path: PathBuf,
}

impl Config {
    pub fn new() -> Result<Self, config::ConfigError> {
        let s = config::Config::builder()
            .add_source(Environment::default().try_parsing(true))
            .build()?;

        // You can deserialize (and thus freeze) the entire configuration as
        s.try_deserialize()
    }
}

pub static GLOBAL_CONFIG: Lazy<Config> = Lazy::new(|| Config::new().unwrap());
