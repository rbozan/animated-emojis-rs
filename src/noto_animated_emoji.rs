use std::{fs::File, io::BufReader, path::PathBuf};

use serde::Deserialize;

#[derive(Debug, PartialEq, Eq, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
struct Emoji {
    base: Vec<u32>,
    alternates: Vec<Vec<u32>>,
    emoticons: Vec<String>,
    shortcodes: Vec<String>,
    animated: bool,
    directional: bool,
}

impl Emoji {
    fn as_hex(&self) -> String {
        self.base
            .iter()
            .map(|base| format!("{:x}", base))
            .collect::<Vec<_>>()
            .join("_")
    }

    fn lottie_download_url(&self) -> String {
        format!(
            "https://fonts.gstatic.com/s/e/notoemoji/latest/{}/lottie.json",
            self.as_hex()
        )
    }
}

#[derive(Debug, PartialEq, Eq, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
struct EmojiGroup {
    group: String,
    emoji: Vec<Emoji>,
}

fn deserialize_noto_animated_emojis_metadata(
    path: &PathBuf,
) -> Result<Vec<EmojiGroup>, Box<dyn std::error::Error>> {
    let file = File::open(path)?;
    let reader = BufReader::new(file);

    let emoji_groups: Vec<EmojiGroup> = serde_json::from_reader(reader)?;

    Ok(emoji_groups)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::GLOBAL_CONFIG;
    use serde_json::json;

    #[test]
    fn it_deserializes_metadata() {
        let path = &GLOBAL_CONFIG.noto_emoji_metadata_path;

        let result = deserialize_noto_animated_emojis_metadata(&path);
        insta::assert_yaml_snapshot!(result.unwrap());
    }

    #[test]
    fn it_returns_correct_lottie_download_url() {
        let emoji: Emoji = serde_json::from_value(json!({
          "base": [
            127987,
            65039
          ],
          "alternates": [],
          "emoticons": [],
          "shortcodes": [
            ":white-flag:"
          ],
          "animated": true,
          "directional": false
        }))
        .unwrap();

        insta::assert_snapshot!(emoji.lottie_download_url());
    }
}
