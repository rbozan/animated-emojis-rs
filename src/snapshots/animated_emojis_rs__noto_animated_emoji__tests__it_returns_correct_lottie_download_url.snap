---
source: src/noto_animated_emoji.rs
expression: emoji.lottie_download_url()
---
https://fonts.gstatic.com/s/e/notoemoji/latest/1f3f3_fe0f/lottie.json
